README.md

```diff
@@ -25,3 +25,109 @@ Hi, I'm your host, Dima ([Github](https://github.com/dkorolev), [LinkedIn](https
+
+   Docker is a must have these days, whether we like it or not. Ideally, each component should be designed such that it can be built & run on bare metal, or packed into a Debian module or into a unikernel of its own, or even be coupled with other components into a single combined piece. Still, for the purposes of keeping the design clean, each component should come with a `Dockerfile`, so that a `docker compose`-based setup can be natively used to spin up, run, use, and test the entire constellation.
+
+   In particular, if a certain component requires warmup, or has nontrivial dependencies, we implement these in a Docker-native way first, using health checks and additional containers so that the dependency graph is respected by any well-designed test.
+
+   I also want our components to be agnostic with respect to now-common standard building blocks. There should be no "Database" or "Cache" or "Message Queue" or "Leader Elections" components, and there should be no Postgres/MySQL/Mongo, Redis/Memcached, Kafka/RabbitMQ, or Consul/etcd ones. Instead, each component should serve a specific well-articulated logical function.
+
+   Of course, certain implementations of certain components would effectively be thin wrappers over standard building blocks such as Postgres or Kafka; by no means we plan to re-implement their functionality from scratch. What is important is that our components are logical building blocks, tailored to our needs, not to what a particular existing solution such as Postgres can provide. And the API contract for our component should highlight its function, such as "transactionally update X", or "broadcast state change Y", not its implementation detail such as "run this SQL query" or "update this Consul DNS record".
+
+   This should be done such that it is possible to implement the same functionality using multiple underlying building blocks, and effectively compare them to one another, instead of fixating on a particular standard component as the "building block of choice". However tempting it may be, no design we provide should be set on using any particular technology as the backbone. In fact, if a particular technology appears to be a perfect fit, then we likely are either not attacking a generic enough problem, or not thinking of it broadly enough.
+
+3. **Contract-Centric**.
+
+   By contract-centric I mostly mean API-centric, but I keep teaching people to reason in terms of contracts, not APIs. Focusing on the specs of APIs of individual components often is about not seeing the forest behind the trees, and The Contract is a far more encompassing term.
+
+   I would like every module to expose its main interfaces via HTTP/JSON first. This enables demonstrating the core functionality of each individual component from the command line right away, which is a good habit to have, and a great design pattern overall.
```

Comment:
Are you planning to store the changelog? I am referring to some sort of ADR mostly for updated contracts.

Reply:
Of course!